crates/cli-support/src/js/mod.rs

```diff
@@ -916,149 +881,49 @@ impl<'a> Context<'a> {
-            String::from(
-                "
-                if (typeof(obj) === 'number') throw new Error('corrupt slab');
-                obj.cnt -= 1;
-                if (obj.cnt > 0) return;
-                ",
-            )
-        } else {
-            String::from(
-                "
-                obj.cnt -= 1;
-                if (obj.cnt > 0) return;
-                ",
-            )
-        };
+        self.expose_global_heap();
+        self.expose_global_heap_next();
         self.global(&format!(
             "
-            function dropRef(idx) {{
-                {}
-                idx = idx >> 1;
+            function dropObject(idx) {{
                 if (idx < {}) return;
-                let obj = slab[idx];
-                {}
-                // If we hit 0 then free up our space in the slab
-                slab[idx] = slab_next;
-                slab_next = idx;
+                heap[idx] = heap_next;
+                heap_next = idx;
             }}
             ",
-            validate_owned,
-            INITIAL_SLAB_VALUES.len(),
-            dec_ref
-        ));
-    }
-
-    fn expose_global_stack(&mut self) {
-        if !self.exposed_globals.insert("stack") {
-            return;
-        }
-        self.global(&format!(
-            "
-            const stack = [];
-        "
+            INITIAL_HEAP_OFFSET + INITIAL_HEAP_VALUES.len(),
         ));
-        if self.config.debug {
-            self.export(
-                "assertStackEmpty",
-                "
-                function() {
-                    if (stack.length === 0) return;
-                    throw new Error('stack is not currently empty');
-                }
-                ",
-                None,
-            );
-        }
     }
 
-    fn expose_global_slab(&mut self) {
-        if !self.exposed_globals.insert("slab") {
+    fn expose_global_heap(&mut self) {
+        if !self.exposed_globals.insert("heap") {
             return;
         }
-        let initial_values = INITIAL_SLAB_VALUES
-            .iter()
-            .map(|s| format!("{{ obj: {} }}", s))
-            .collect::<Vec<_>>();
-        self.global(&format!("const slab = [{}];", initial_values.join(", ")));
-        if self.config.debug {
-            self.export(
-                "assertSlabEmpty",
-                &format!(
-                    "
-                    function() {{
-                        for (let i = {}; i < slab.length; i++) {{
-                            if (typeof(slab[i]) === 'number') continue;
-                            throw new Error('slab is not currently empty');
-                        }}
-                    }}
-                    ",
-                    initial_values.len()
-                ),
-                None,
-            );
+        self.global("const heap = [];");
+        let init = format!(
+            "for (let i = 0; i < {}; i++) heap.push(undefined);",
+            INITIAL_HEAP_OFFSET,
```

Comment:
Slightly better:

```js
const heap = new Array(INITIAL_HEAP_OFFSET);
heap.fill(undefined);
```

Reply:
Oh nice! I'll omit the `fill` as I don't think we actually need it, and otherwise this works great